### DISCUSSION TITLE
[@types/chrome] types of NotificationOptions is error

### DISCUSSION THREAD

**Discussion by wfk007** [DefinitelyTyped/DefinitelyTyped#57548] [Issues with a @types package]
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9551b738657938d8d1725dc9c329c227fa27273d/types/chrome/index.d.ts#L5417

why `?` with `undefined`
```ts
{
  type?: string | undefined;
}
```

is `type` and `iconUrl` partial？but I got
> Unchecked runtime.lastError: Some of the required properties are missing: type, iconUrl, title and message.

is `type` string? but I got
> Uncaught TypeError: Error in invocation of notifications.create(optional string notificationId, notifications.NotificationOptions options, optional function callback): Error at parameter 'options': Error at property 'type': Value must be one of basic, image, list, progress.

[a comment by an automated account is not shown]

**Comment by nicolas377** [comment]
Hey @wfk007! It seems there are two questions here, I'll try my best to answer both:

> why ? with undefined

Typescript 4.4 added a new option, exactOptionalPropertyTypes, that changed the way optional properties were handled. Because of that, all the optional properties had undefined added onto them in one massive PR (#54317)

> is type and iconUrl partial?

No. Anything partial would be run through the Partial type, and wouldn't be an interface.

As for your errors, both of them seem to be occurring at runtime, but there's no code for us to check. Could you provide the code that's being referenced in the runtime errors?

**Reply by wfk007** [reply]
> Hey @wfk007! It seems there are two questions here, I'll try my best to answer both:
> 
> > why ? with undefined
> 
> Typescript 4.4 added a new option, exactOptionalPropertyTypes, that changed the way optional properties were handled. Because of that, all the optional properties had undefined added onto them in one massive PR (#54317)
> 
> > is type and iconUrl partial?
> 
> No. Anything partial would be run through the Partial type, and wouldn't be an interface.
> 
> As for your errors, both of them seem to be occurring at runtime, but there's no code for us to check. Could you provide the code that's being referenced in the runtime errors?

I added a repository：https://github.com/wfk007/test-chrome-extension
steps：
1. npm install（to install pkg @types/chrome）
2. load this extension in chrome brwoser
3. update code in background.js（you can follow the step in chrome.notifications.create()）
4. view logs in console panel in service worker

**Reply by wfk007** [reply]
thx reply, I got why `?` with `undefined `

**Reply by nicolas377** [reply]
This doesn't seem to be a problem with types, since you're using .js files, and you're getting runtime errors instead of compile-time errors.

As for your runtime errors, I'd point you to the chrome docs [here](https://developer.chrome.com/docs/extensions/reference/notifications/#type-NotificationOptions). Some of the options are marked with `Required for notifications.create method.` Also, in your code, the `type` parameter is defined multiple times, causing another error.

**Reply by wfk007** [reply]
> This doesn't seem to be a problem with types

ARE YOU SERIOUS？DO YOU REALLY UNDERSTAND WHAT I AM TALKING ABOUT?
Even if you write with `ts` file, you will STILL NOT get ANY TYPE ERROR during compilation with @type/chrome and STILL get runtime errors!  Because I was busy doing other things in the day time, I just supply a simple javascript demo. Just to inform you of the meaning

> Some of the options are marked with Required for notifications.create method.

some options in docs are marked as Required, But WHY your types NOT? IS NOT IT ERROR?

> Also, in your code, the type parameter is defined multiple times, causing another error.

type parameter is defined multiple times IS to show you different cases, it is a procedure of thinking when writing code, hard to understand?

**Reply by nicolas377** [reply]
I've put up a PR (#57602) to enforce the needed options. It might get held up for a few days (hopefully not), since there aren't many other active maintainers. Thanks for pointing this out!

**Reply by wfk007** [reply]
thx pr, I think using template literal types would be much better
```ts
{
    type: 'basic' | 'image' | 'list' | 'progress' | (string & { _?: never });
}
```

**Reply by nicolas377** [reply]
Version 0.0.169 is live with the fix to this.

**Reply by wfk007** [reply]
thanks for your great work
